File: sdk/python/inkbox/contacts/resources/contacts.py
"""
inkbox/contacts/resources/contacts.py

Contacts CRUD + search + lookup.
"""

from __future__ import annotations

from datetime import date
from typing import TYPE_CHECKING, Any, Literal
from uuid import UUID

from inkbox.contacts.resources.contact_access import ContactAccessResource
from inkbox.contacts.resources.contact_facts import ContactFactsResource
from inkbox.contacts.resources.correspondence import ContactCorrespondenceResource
from inkbox.contacts.resources.vcards import VCardsResource
from inkbox.contacts.types import (
    Contact,
    ContactAddress,
    ContactBulkDeleteResult,
    ContactCustomField,
    ContactDate,
    ContactEmail,
    ContactPhone,
    ContactReviewStatus,
    ContactWebsite,
)

if TYPE_CHECKING:
    from inkbox._http import HttpTransport

_BASE = "/contacts"
_UNSET = object()


def _items_to_wire(items: list[Any] | None) -> list[dict[str, Any]] | None:
    if items is None:
        return None
    return [item.to_wire() if hasattr(item, "to_wire") else item for item in items]


class ContactsResource:
    """Organization-wide contacts and contact memory."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http
        self._access = ContactAccessResource(http)
        self._facts = ContactFactsResource(http)
        self._correspondence = ContactCorrespondenceResource(http)
        self._vcards = VCardsResource(http)

    @property
    def access(self) -> ContactAccessResource:
        """Compatibility access records (read-only)."""
        return self._access

    @property
    def facts(self) -> ContactFactsResource:
        """Facts recorded for contacts."""
        return self._facts

    @property
    def correspondence(self) -> ContactCorrespondenceResource:
        """Correspondence across contact channels."""
        return self._correspondence

    @property
    def vcards(self) -> VCardsResource:
        """vCard import / export."""
        return self._vcards

    def list(
        self,
        *,
        q: str | None = None,
        order: Literal["name", "recent"] | str | None = None,
        limit: int | None = None,
        offset: int | None = None,
        review_status: list[ContactReviewStatus | str] | None = None,
    ) -> list[Contact]:
        """List contacts with optional substring search.

        Args:
            q: Optional case-insensitive substring filter across
                ``preferred_name``, ``given_name``, ``family_name``,
                ``company_name``, ``job_title``, and ``notes``. Max 100 chars.
            order: ``"name"`` or ``"recent"`` sort order (server default applies).
            limit: Max rows to return.
            offset: Offset for paging, from 0 through 10,000.
            review_status: Review states to include.
        """
        params: dict[str, Any] = {}
        if q is not None:
            params["q"] = q
        if order is not None:
            params["order"] = order
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        if review_status is not None:
            params["review_status"] = [str(status) for status in review_status]
        data = self._http.get(_BASE, params=params)
        items = data["items"] if isinstance(data, dict) and "items" in data else data
        return [Contact._from_dict(c) for c in items]

    def lookup(
        self,
        *,
        email: str | None = None,
        email_contains: str | None = None,
        email_domain: str | None = None,
        phone: str | None = None,
        phone_contains: str | None = None,
    ) -> list[Contact]:
        """Reverse-lookup contacts by a single field.

        Exactly one of the five arguments must be supplied; passing zero or
        more than one raises ``ValueError`` before hitting the server.
        """
        supplied = {
            "email": email,
            "email_contains": email_contains,
            "email_domain": email_domain,
            "phone": phone,
            "phone_contains": phone_contains,
        }
        non_none = {k: v for k, v in supplied.items() if v is not None}
        if len(non_none) != 1:
            raise ValueError(
                "lookup() requires exactly one of: email, email_contains, "
                "email_domain, phone, phone_contains.",
            )
        data = self._http.get(f"{_BASE}/lookup", params=non_none)
        items = data["items"] if isinstance(data, dict) and "items" in data else data
        return [Contact._from_dict(c) for c in items]

    def get(self, contact_id: UUID | str) -> Contact:
        """Fetch a single contact by id."""
        data = self._http.get(f"{_BASE}/{contact_id}")
        return Contact._from_dict(data)

    def create(
        self,
        *,
        preferred_name: str | None = None,
        name_prefix: str | None = None,
        given_name: str | None = None,
        middle_name: str | None = None,
        family_name: str | None = None,
        name_suffix: str | None = None,
        company_name: str | None = None,
        job_title: str | None = None,
        birthday: date | str | None = None,
        notes: str | None = None,
        emails: list[ContactEmail] | None = None,
        phones: list[ContactPhone] | None = None,
        websites: list[ContactWebsite] | None = None,
        dates: list[ContactDate] | None = None,
        addresses: list[ContactAddress] | None = None,
        custom_fields: list[ContactCustomField] | None = None,
    ) -> Contact:
        """Create a new contact.

        Args:
            birthday: Optional ISO date (``YYYY-MM-DD``) or :class:`datetime.date`.
        """
        body: dict[str, Any] = {}
        for name, value in (
            ("preferred_name", preferred_name),
            ("name_prefix", name_prefix),
            ("given_name", given_name),
            ("middle_name", middle_name),
            ("family_name", family_name),
            ("name_suffix", name_suffix),
            ("company_name", company_name),
            ("job_title", job_title),
            ("notes", notes),
        ):
            if value is not None:
                body[name] = value
        if birthday is not None:
            body["birthday"] = (
                birthday.isoformat() if isinstance(birthday, date) else birthday
            )
        for name, items in (
            ("emails", emails),
            ("phones", phones),
            ("websites", websites),
            ("dates", dates),
            ("addresses", addresses),
            ("custom_fields", custom_fields),
        ):
            wire = _items_to_wire(items)
            if wire is not None:
                body[name] = wire
        data = self._http.post(_BASE, json=body)
        return Contact._from_dict(data)

    def update(
        self,
        contact_id: UUID | str,
        *,
        preferred_name: str | None = _UNSET,  # type: ignore[assignment]
        name_prefix: str | None = _UNSET,  # type: ignore[assignment]
        given_name: str | None = _UNSET,  # type: ignore[assignment]
        middle_name: str | None = _UNSET,  # type: ignore[assignment]
        family_name: str | None = _UNSET,  # type: ignore[assignment]
        name_suffix: str | None = _UNSET,  # type: ignore[assignment]
        company_name: str | None = _UNSET,  # type: ignore[assignment]
        job_title: str | None = _UNSET,  # type: ignore[assignment]
        birthday: date | str | None = _UNSET,  # type: ignore[assignment]
        notes: str | None = _UNSET,  # type: ignore[assignment]
        emails: list[ContactEmail] | None = _UNSET,  # type: ignore[assignment]
        phones: list[ContactPhone] | None = _UNSET,  # type: ignore[assignment]
        websites: list[ContactWebsite] | None = _UNSET,  # type: ignore[assignment]
        dates: list[ContactDate] | None = _UNSET,  # type: ignore[assignment]
        addresses: list[ContactAddress] | None = _UNSET,  # type: ignore[assignment]
        custom_fields: list[ContactCustomField] | None = _UNSET,  # type: ignore[assignment]
        review_status: ContactReviewStatus | str = _UNSET,  # type: ignore[assignment]
    ) -> Contact:
        """JSON-merge-patch update.

        Only provided fields are sent; omit a field to leave it unchanged.
        Pass a scalar as ``None`` to clear it.
        """
        body: dict[str, Any] = {}
        scalar_fields = (
            ("preferred_name", preferred_name),
            ("name_prefix", name_prefix),
            ("given_name", given_name),
            ("middle_name", middle_name),
            ("family_name", family_name),
            ("name_suffix", name_suffix),
            ("company_name", company_name),
            ("job_title", job_title),
            ("notes", notes),
        )
        for name, value in scalar_fields:
            if value is not _UNSET:
                body[name] = value
        if review_status is not _UNSET:
            body["review_status"] = review_status
        if birthday is not _UNSET:
            body["birthday"] = (
                birthday.isoformat() if isinstance(birthday, date) else birthday
            )
        list_fields = (
            ("emails", emails),
            ("phones", phones),
            ("websites", websites),
            ("dates", dates),
            ("addresses", addresses),
            ("custom_fields", custom_fields),
        )
        for name, items in list_fields:
            if items is _UNSET:
                continue
            body[name] = _items_to_wire(items) if items is not None else None
        data = self._http.patch(f"{_BASE}/{contact_id}", json=body)
        return Contact._from_dict(data)

    def merge(
        self,
        contact_id: UUID | str,
        *,
        losing_contact_ids: list[UUID | str],
        field_sources: dict[str, UUID | str] | None = None,
    ) -> Contact:
        """Merge contacts into the contact identified by ``contact_id``."""
        body = {
            "losing_contact_ids": [str(value) for value in losing_contact_ids],
            "field_sources": {
                name: str(value) for name, value in (field_sources or {}).items()
            },
        }
        data = self._http.post(f"{_BASE}/{contact_id}/merge", json=body)
        return Contact._from_dict(data)

    def delete(self, contact_id: UUID | str) -> None:
        """Delete a contact."""
        self._http.delete(f"{_BASE}/{contact_id}")

    def bulk_delete(
        self,
        contact_ids: list[UUID | str],
    ) -> ContactBulkDeleteResult:
        """Delete multiple contacts and return per-contact outcomes."""
        data = self._http.post(
            f"{_BASE}/bulk-delete",
            json={"contact_ids": [str(contact_id) for contact_id in contact_ids]},
        )
        return ContactBulkDeleteResult._from_dict(data)

File: cli/src/commands/contacts.ts
import { Command, Option } from "commander";
import { readFileSync, writeFileSync } from "node:fs";
import type {
  ContactFactKind,
  ContactReviewStatus,
  CorrespondenceChannel,
  CorrespondenceContentMode,
  CorrespondenceOrder,
  CorrespondenceTranscriptMode,
  CreateContactOptions,
  MergeContactsOptions,
  ReplaceContactCommunicationPolicy,
  UpdateContactPermissions,
} from "@inkbox/sdk";
import { createClient, getGlobalOpts } from "../client.js";
import { output } from "../output.js";
import { withErrorHandler } from "../errors.js";
import { parsePolicyPagination } from "../pagination.js";

function parseJsonArg<T>(raw: string, label: string): T {
  try {
    return JSON.parse(raw) as T;
  } catch (err) {
    throw new Error(
      `Could not parse ${label} as JSON: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

function collectValues(value: string, previous: string[] = []): string[] {
  return [
    ...previous,
    ...value.split(",").map((part) => part.trim()).filter(Boolean),
  ];
}

/** Validate objects before serialization can discard unrecognized settings. */
function policyObject(value: unknown, fields: string[], path: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new Error(`${path} must be an object`);
  const record = value as Record<string, unknown>;
  const unknown = Object.keys(record).filter((key) => !fields.includes(key));
  if (unknown.length) throw new Error(`Unknown field in ${path}: ${unknown.join(", ")}`);
  return record;
}

/** Validate a complete visibility portion. */
function validatePolicyPortion(defaults: unknown, identities: unknown, fields: string[], path: string): void {
  const validateDecisions = (row: Record<string, unknown>, label: string): void => {
    for (const key of fields) {
      if (!["inherit", "allow", "block"].includes(row[key] as string)) throw new Error(`${label}.${key} must be inherit, allow, or block`);
    }
  };
  validateDecisions(policyObject(defaults, fields, `${path}.defaults`), `${path}.defaults`);
  if (!Array.isArray(identities)) throw new Error(`${path}.identities must be an array`);
  if (identities.length > 500) throw new Error(`${path}.identities cannot exceed 500 entries`);
  const seen = new Set<string>();
  for (const [index, value] of identities.entries()) {
    const label = `${path}.identities[${index}]`;
    const row = policyObject(value, ["identityId", ...fields], label);
    if (typeof row.identityId !== "string" || !row.identityId) throw new Error(`${label}.identityId is required`);
    if (seen.has(row.identityId)) throw new Error(`${path}.identities contains duplicate identities`);
    seen.add(row.identityId);
    validateDecisions(row, label);
  }
}

/** Parse a policy file while preserving omission of visibility settings. */
export function parseContactPolicyFile(raw: string): ReplaceContactCommunicationPolicy {
  const body = policyObject(parseJsonArg<unknown>(raw, "policy file"), ["expectedRevision", "identityId", "addresses", "visibility"], "policy");
  if (!Number.isSafeInteger(body.expectedRevision) || (body.expectedRevision as number) < 0) throw new Error("policy.expectedRevision must be a nonnegative integer");
  if (typeof body.identityId !== "string" || !body.identityId) throw new Error("policy.identityId is required");
  if (!Array.isArray(body.addresses) || body.addresses.length > 500) throw new Error("policy.addresses must be an array of at most 500 entries");
  for (const [index, value] of body.addresses.entries()) {
    const label = `policy.addresses[${index}]`;
    const row = policyObject(value, ["kind", "value", "action", "expectedAction"], label);
    if (!["email", "phone"].includes(row.kind as string)) throw new Error(`${label}.kind must be email or phone`);
    if (typeof row.value !== "string" || !row.value) throw new Error(`${label}.value is required`);
    for (const key of ["action", "expectedAction"]) {
      if (!["inherit", "allow", "block"].includes(row[key] as string)) throw new Error(`${label}.${key} must be inherit, allow, or block`);
    }
  }
  if (Object.hasOwn(body, "visibility")) {
    const visibility = policyObject(body.visibility, ["defaults", "identities"], "policy.visibility");
    validatePolicyPortion(visibility.defaults, visibility.identities, ["profile", "memories"], "policy.visibility");
  }
  return body as unknown as ReplaceContactCommunicationPolicy;
}

export function parseContactPermissionsFile(raw: string): UpdateContactPermissions {
  const body = policyObject(parseJsonArg<unknown>(raw, "permissions file"), ["emails", "phones", "profile", "memories"], "permissions");
  for (const key of ["emails", "phones"]) {
    if (!Object.hasOwn(body, key)) continue;
    const value = body[key];
    if (value === null || typeof value !== "object" || Array.isArray(value)) throw new Error(`permissions.${key} must be an address-to-boolean object`);
    if (Object.keys(value).length > 50) throw new Error(`permissions.${key} cannot exceed 50 entries`);
    for (const allowed of Object.values(value)) {
      if (typeof allowed !== "boolean") throw new Error(`permissions.${key} values must be true or false`);
    }
  }
  for (const key of ["profile", "memories"]) {
    if (Object.hasOwn(body, key) && typeof body[key] !== "boolean") throw new Error(`permissions.${key} must be true or false`);
  }
  return body as UpdateContactPermissions;
}

function registerContactsAccessCommands(parent: Command): void {
  const permissions = parent.command("permissions").description("Selected-agent yes/no contact access (admin credentials)");
  permissions.command("get <handle> <contact-id>").description("Read effective email, phone, profile, and memory access")
    .action(withErrorHandler(async function (this: Command, handle: string, contactId: string): Promise<void> {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).contacts.permissions.get(handle, contactId) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));
  permissions.command("set <handle> <contact-id>").description("Save yes/no choices; omitted fields and addresses stay unchanged")
    .requiredOption("--file <path>", "JSON file with optional emails/phones boolean maps and profile/memories booleans")
    .action(withErrorHandler(async function (this: Command, handle: string, contactId: string, options: { file: string }): Promise<void> {
      const opts = getGlobalOpts(this);
      const body = parseContactPermissionsFile(readFileSync(options.file, "utf8"));
      output(await createClient(opts).contacts.permissions.update(handle, contactId, body) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));

  const policy = parent.command("communication-policy").description("Advanced address policies, visibility settings, and previews");
  policy.command("list-management <handle>").description("Manage contact permissions, including hidden contacts (admin credentials)")
    .option("--q <query>", "Search contact details")
    .option("--limit <number>", "Page size", "50").option("--offset <number>", "Page offset", "0")
    .addOption(new Option("--order <order>", "Sort order").choices(["name", "recent"]).default("recent"))
    .addOption(new Option("--review-status <status...>", "Contact review states").choices(["confirmed", "unreviewed"]))
    .action(withErrorHandler(async function (this: Command, handle: string, options: {
      q?: string; limit: string; offset: string; order: "name" | "recent"; reviewStatus?: ContactReviewStatus[];
    }): Promise<void> {
      const opts = getGlobalOpts(this);
      const page = await createClient(opts).contacts.communicationPolicy.listManagementForIdentity(handle, {
        ...options, ...parsePolicyPagination(options),
      });
      output(opts.json ? page : page.items.map((row) => ({ id: row.contact.id, contact: row.contact.preferredName,
        ...row.effective, revision: row.revision })),
      { json: !!opts.json, columns: ["id", "contact", "email", "phone", "profile", "memories", "revision"] });
    }));
  policy.command("get <contact-id>").description("Read a contact policy (admin credentials)")
    .option("--identity-id <uuid>", "Include this agent's address choices and effective visibility")
    .action(withErrorHandler(async function (this: Command, contactId: string, options: { identityId?: string }): Promise<void> {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).contacts.communicationPolicy.get(contactId, options.identityId) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));
  policy.command("set <contact-id>").description("Replace a contact policy (admin credentials)")
    .requiredOption("--file <path>", "JSON file with expectedRevision, identityId, addresses, and optional visibility")
    .action(withErrorHandler(async function (this: Command, contactId: string, options: { file: string }): Promise<void> {
      const opts = getGlobalOpts(this);
      const body = parseContactPolicyFile(readFileSync(options.file, "utf8"));
      output(await createClient(opts).contacts.communicationPolicy.replace(contactId, body) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));
  policy.command("preview <contact-id> <identity-id>").description("Preview contact visibility (admin credentials)")
    .action(withErrorHandler(async function (this: Command, contactId: string, identityId: string): Promise<void> {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).contacts.communicationPolicy.preview(contactId, identityId) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));
  policy.command("list <handle>").description("List an identity's visible contact permissions")
    .option("--limit <number>", "Page size", "50").option("--offset <number>", "Page offset", "0")
    .action(withErrorHandler(async function (this: Command, handle: string, options: { limit: string; offset: string }): Promise<void> {
      const opts = getGlobalOpts(this);
      output(await createClient(opts).contacts.communicationPolicy.listForIdentity(handle, parsePolicyPagination(options)) as unknown as Record<string, unknown>, { json: !!opts.json });
    }));

  const access = parent
    .command("access")
    .description("Compatibility access view");

  access
    .command("list <contact-id>")
    .description("List compatibility access rows on a contact")
    .action(
      withErrorHandler(async function (this: Command, contactId: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const rows = await inkbox.contacts.access.list(contactId);
        output(rows, {
          json: !!opts.json,
          columns: ["id", "contactId", "identityId", "createdAt"],
        });
      }),
    );

}

const FACT_KINDS: ContactFactKind[] = ["profile", "preference", "context"];

function parseFactKind(value: string): ContactFactKind {
  if (!(FACT_KINDS as string[]).includes(value)) {
    throw new Error(`--kind must be one of: ${FACT_KINDS.join(", ")}`);
  }
  return value as ContactFactKind;
}

function registerContactFactsCommands(parent: Command): void {
  const facts = parent.command("facts").description("Contact memory facts and citations");

  facts
    .command("list <contact-id>")
    .description("List a contact's facts")
    .option(
      "--include-expired",
      "Also list expired context facts (locked facts remain active)",
    )
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: { includeExpired?: boolean },
      ) {
        const opts = getGlobalOpts(this);
        const rows = await createClient(opts).contacts.facts.list(contactId, {
          includeExpired: !!cmdOpts.includeExpired,
        });
        output(rows, {
          json: !!opts.json,
          columns: ["id", "kind", "content", "origin", "confidence", "expiresAt", "updatedAt"],
        });
      }),
    );

  facts
    .command("get <contact-id> <fact-id>")
    .description("Fetch a contact fact")
    .action(
      withErrorHandler(async function (this: Command, contactId: string, factId: string) {
        const opts = getGlobalOpts(this);
        const fact = await createClient(opts).contacts.facts.get(contactId, factId);
        output(fact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  facts
    .command("citation <contact-id> <fact-id> <citation-id>")
    .description("Resolve a fact citation")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        factId: string,
        citationId: string,
      ) {
        const opts = getGlobalOpts(this);
        const citation = await createClient(opts).contacts.facts.resolveCitation(
          contactId,
          factId,
          citationId,
        );
        output(citation as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  facts
    .command("citation-url <source-url>")
    .description("Resolve an available fact citation URL")
    .action(
      withErrorHandler(async function (this: Command, sourceUrl: string) {
        const opts = getGlobalOpts(this);
        const citation = await createClient(opts).contacts.facts.resolveCitationUrl(sourceUrl);
        output(citation as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  facts
    .command("create <contact-id>")
    .description(
      "Record a fact by hand (admin-scoped API key required; hand-written facts never expire)",
    )
    .requiredOption("--content <text>", "What to remember about the contact")
    .requiredOption("--kind <kind>", "profile, preference, or context")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: { content: string; kind: string },
      ) {
        const opts = getGlobalOpts(this);
        const fact = await createClient(opts).contacts.facts.create(contactId, {
          content: cmdOpts.content,
          kind: parseFactKind(cmdOpts.kind),
        });
        output(fact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  facts
    .command("update <contact-id> <fact-id>")
    .description(
      "Edit a fact (admin-scoped API key required; any edit makes it manually maintained and revives it; content changes remove citations)",
    )
    .option("--content <text>", "Replacement content")
    .option("--kind <kind>", "profile, preference, or context")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        factId: string,
        cmdOpts: { content?: string; kind?: string },
      ) {
        const opts = getGlobalOpts(this);
        if (cmdOpts.content === undefined && cmdOpts.kind === undefined) {
          throw new Error("Pass --content, --kind, or both");
        }
        const fact = await createClient(opts).contacts.facts.update(contactId, factId, {
          content: cmdOpts.content,
          kind: cmdOpts.kind === undefined ? undefined : parseFactKind(cmdOpts.kind),
        });
        output(fact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  facts
    .command("delete <contact-id> <fact-id>")
    .description("Delete a contact fact (admin-scoped API key required)")
    .action(
      withErrorHandler(async function (this: Command, contactId: string, factId: string) {
        const opts = getGlobalOpts(this);
        const result = await createClient(opts).contacts.facts.delete(contactId, factId);
        output(result as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );
}

export function registerContactsCommands(program: Command): void {
  const contacts = program
    .command("contacts")
    .description("Org-wide contacts, memory, correspondence, and vCard");

  contacts
    .command("list")
    .description("List contacts")
    .option("--q <query>", "Case-insensitive substring search (≤100 chars)")
    .option("--order <order>", "name or recent")
    .option(
      "--review-status <status>",
      "Filter by confirmed or unreviewed (repeat or comma-separate)",
      collectValues,
    )
    .option("--limit <n>", "Max rows", (v) => parseInt(v, 10))
    .option("--offset <n>", "Offset (0-10000)", (v) => parseInt(v, 10))
    .action(
      withErrorHandler(async function (
        this: Command,
        cmdOpts: {
          q?: string;
          order?: string;
          reviewStatus?: string[];
          limit?: number;
          offset?: number;
        },
      ) {
        const opts = getGlobalOpts(this);
        const rows = await createClient(opts).contacts.list({
          ...cmdOpts,
          reviewStatus: cmdOpts.reviewStatus as ContactReviewStatus[] | undefined,
        });
        output(rows, {
          json: !!opts.json,
          columns: [
            "id",
            "preferredName",
            "givenName",
            "familyName",
            "companyName",
            "jobTitle",
            "memoryCount",
            "updatedAt",
          ],
        });
      }),
    );

  contacts
    .command("get <contact-id>")
    .description("Fetch a single contact")
    .action(
      withErrorHandler(async function (this: Command, contactId: string) {
        const opts = getGlobalOpts(this);
        const contact = await createClient(opts).contacts.get(contactId);
        output(contact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("correspondence <contact-id>")
    .description("List correspondence with a contact across channels")
    .option("-i, --identity <uuid>", "Agent identity UUID")
    .option("--channels <channel>", "Channels to include (repeat or comma-separate)", collectValues)
    .option("--after <datetime>", "Only items at or after this time")
    .option("--before <datetime>", "Only items before this time")
    .option("--limit-per-channel <n>", "Default limit per channel", (v) => parseInt(v, 10))
    .option("--email-limit <n>", "Email limit", (v) => parseInt(v, 10))
    .option("--sms-limit <n>", "SMS limit", (v) => parseInt(v, 10))
    .option("--imessage-limit <n>", "iMessage limit", (v) => parseInt(v, 10))
    .option("--calls-limit <n>", "Call limit", (v) => parseInt(v, 10))
    .option("--cursor <cursor>", "Pagination cursor")
    .option("--order <order>", "asc or desc")
    .option("--content <mode>", "metadata, preview, or full")
    .option("--transcripts <mode>", "none, abridged, or full")
    .option("--include-failed", "Include failed correspondence", false)
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: {
          identity?: string;
          channels?: string[];
          after?: string;
          before?: string;
          limitPerChannel?: number;
          emailLimit?: number;
          smsLimit?: number;
          imessageLimit?: number;
          callsLimit?: number;
          cursor?: string;
          order?: string;
          content?: string;
          transcripts?: string;
          includeFailed?: boolean;
        },
      ) {
        const opts = getGlobalOpts(this);
        const { identity, ...correspondenceOptions } = cmdOpts;
        const result = await createClient(opts).contacts.correspondence.get(contactId, {
          ...correspondenceOptions,
          channels: cmdOpts.channels as CorrespondenceChannel[] | undefined,
          order: cmdOpts.order as CorrespondenceOrder | undefined,
          content: cmdOpts.content as CorrespondenceContentMode | undefined,
          transcripts: cmdOpts.transcripts as CorrespondenceTranscriptMode | undefined,
          identityId: identity,
        });
        output(result as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("merge <contact-id>")
    .description(
      "Merge contacts (admin-scoped key required; rejected when a memory kind or total goes over; delete from named kinds, or any active fact for total)",
    )
    .requiredOption("--losing <contact-id...>", "Contact IDs to merge into the survivor")
    .option("--field-sources <json>", "JSON object mapping profile fields to source contact IDs")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: { losing: string[]; fieldSources?: string },
      ) {
        const opts = getGlobalOpts(this);
        const fieldSources = cmdOpts.fieldSources
          ? parseJsonArg<MergeContactsOptions["fieldSources"]>(
              cmdOpts.fieldSources,
              "--field-sources",
            )
          : undefined;
        const result = await createClient(opts).contacts.merge(contactId, {
          losingContactIds: cmdOpts.losing,
          fieldSources,
        });
        output(result as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("create")
    .description("Create a contact (pass the full payload as JSON)")
    .requiredOption("--json <payload>", "JSON payload matching CreateContactOptions; optional permissions require an admin API key")
    .action(
      withErrorHandler(async function (
        this: Command,
        cmdOpts: { json: string },
      ) {
        const opts = getGlobalOpts(this);
        const payload = parseJsonArg<CreateContactOptions>(cmdOpts.json, "--json payload");
        const inkbox = createClient(opts);
        const contact = await inkbox.contacts.create(payload);
        output(contact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("update <contact-id>")
    .description("JSON-merge-patch update (pass the patch as JSON)")
    .requiredOption("--json <payload>", "JSON patch matching UpdateContactOptions")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: { json: string },
      ) {
        const opts = getGlobalOpts(this);
        const patch = parseJsonArg<Record<string, unknown>>(cmdOpts.json, "--json patch");
        const inkbox = createClient(opts);
        const contact = await inkbox.contacts.update(contactId, patch);
        output(contact as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("delete <contact-id>")
    .description("Delete a contact")
    .action(
      withErrorHandler(async function (this: Command, contactId: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        await inkbox.contacts.delete(contactId);
        console.log(`Deleted contact ${contactId}.`);
      }),
    );

  contacts
    .command("bulk-delete <contact-id...>")
    .description("Delete multiple contacts")
    .action(
      withErrorHandler(async function (this: Command, contactIds: string[]) {
        const opts = getGlobalOpts(this);
        const result = await createClient(opts).contacts.bulkDelete(contactIds);
        output(result as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("lookup")
    .description("Reverse-lookup contacts (exactly one filter required)")
    .option("--email <email>", "Exact email match")
    .option("--email-contains <substr>", "Case-insensitive email substring")
    .option("--email-domain <domain>", "Domain-part match")
    .option("--phone <e164>", "Exact phone match")
    .option("--phone-contains <substr>", "Phone substring")
    .action(
      withErrorHandler(async function (
        this: Command,
        cmdOpts: {
          email?: string;
          emailContains?: string;
          emailDomain?: string;
          phone?: string;
          phoneContains?: string;
        },
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const rows = await inkbox.contacts.lookup(cmdOpts);
        output(rows, {
          json: !!opts.json,
          columns: ["id", "preferredName", "givenName", "familyName", "updatedAt"],
        });
      }),
    );

  contacts
    .command("import <file>")
    .description("Bulk vCard import (text/vcard, ≤5 MiB, ≤1000 cards)")
    .action(
      withErrorHandler(async function (this: Command, file: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const body = readFileSync(file, "utf8");
        const result = await inkbox.contacts.vcards.import(body, "text/vcard");
        output(result as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  contacts
    .command("export-many <contact-id...>")
    .description("Export up to 25 contacts as one vCard document")
    .option("--out <file>", "Write to file instead of stdout")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactIds: string[],
        cmdOpts: { out?: string },
      ) {
        const opts = getGlobalOpts(this);
        const result = await createClient(opts).contacts.vcards.exportMany(contactIds);
        if (cmdOpts.out) {
          writeFileSync(cmdOpts.out, result.vcard, "utf8");
          if (!opts.json) console.log(`Wrote ${cmdOpts.out}`);
        } else if (opts.json) {
          output(result as unknown as Record<string, unknown>, { json: true });
        } else {
          process.stdout.write(result.vcard);
        }
      }),
    );

  contacts
    .command("export <contact-id>")
    .description("Export a single contact as vCard 4.0")
    .option("--out <file>", "Write to file instead of stdout")
    .action(
      withErrorHandler(async function (
        this: Command,
        contactId: string,
        cmdOpts: { out?: string },
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const vcf = await inkbox.contacts.vcards.export(contactId);
        if (cmdOpts.out) {
          writeFileSync(cmdOpts.out, vcf, "utf8");
          if (!opts.json) console.log(`Wrote ${cmdOpts.out}`);
        } else {
          if (opts.json) {
            output({ vcard: vcf }, { json: true });
          } else {
            process.stdout.write(vcf);
          }
        }
      }),
    );

  registerContactsAccessCommands(contacts);
  registerContactFactsCommands(contacts);
}
